assets/src/components/site-scan-context-provider/index.js

```diff
@@ -307,15 +348,41 @@ export function SiteScanContextProvider( {
 	 * Scan site URLs sequentially.
 	 */
 	useEffect( () => {
-		( async () => {
-			if ( status !== STATUS_IDLE ) {
-				return;
+		if ( ! [ STATUS_IDLE, STATUS_IN_PROGRESS ].includes( status ) ) {
+			return;
+		}
+
+		/**
+		 * If there are no more URLs to scan and no URLs are scanned at the
+		 * moment, finish the site scan.
+		 */
+		if ( urlIndexesPendingScan.length === 0 ) {
+			if ( currentlyScannedUrlIndexes.length === 0 ) {
+				dispatch( { type: ACTION_SCAN_COMPLETE } );
 			}
 
-			dispatch( { type: ACTION_SCAN_VALIDATE_URL } );
+			return;
+		}
+
+		if ( shouldDelayValidationRequest || currentlyScannedUrlIndexes.length >= CONCURRENT_VALIDATION_REQUESTS_MAX_COUNT ) {
+			return;
+		}
+
+		setShouldDelayValidationRequest( true );
+
+		const currentlyScannedUrlIndex = urlIndexesPendingScan.shift();
```

Comment:
Question: Was `urlIndexesPendingScan` cloned from the state? Or is this mutation be silently updated in the state? Does that matter?

Reply:
@westonruter This is a very good question and a great catch - thank you for that!

> Was `urlIndexesPendingScan` cloned from the state?
No, it was not. I was wrongly under the impression that it is.

> Or is this mutation be silently updated in the state? Does that matter?
Since a new state should always be calculated in the reducer, mutating the state elsewhere is incorrect. Luckily, this mistake didn't cause any harm since we've been [filtering out an inexistent array index in the reducer](https://github.com/ampproject/amp-wp/pull/6705/files#diff-bcbad85a0f6d8eba326a1158103c8e7e65c028ea7a0ec2c3e2496b983c65fc53R137).

Anyway, I've fixed this issue in [another PR](https://github.com/ampproject/amp-wp/pull/6698/commits/253e5fde42dbe52964dc421ad64baf3d4bd90df2) so that the new state is calculated inside the reducer.